source/text/typesetter.ts

```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
+
+        const upperRight = vec2.create();
+        vec2.add(upperRight, glyph.subTextureOrigin, glyph.subTextureExtent);
+
+        vertex.uvRect[2] = upperRight[0];
+        vertex.uvRect[3] = upperRight[1];
     }
 
+
     /**
-     * Accumulate glyph advances (including kerning) up to the next delimiter or line feed occurrence starting at
-     * the given index for the given label.
-     * @param label - Label to compute accumulated advances for.
-     * @param begin - Index to start glyph advance accumulation at.
-     * @param width - Out parameter: the accumulated width up to the next delimiter (reset to 0).
-     * @returns - The index of the last character that was included in the forward accumulation.
-     * @todo Perhaps switch to another approach, i.e., calculate the all advances between delimiters once and create
-     * an index based lookup...
+     * Compute an initial line anchor w.r.t. the targeted anchoring.
+     * @param label - Label to adjust the y-positions for.
      */
-    protected static forward(label: Label, begin: number, width: number): number {
-        let index: number = begin;
-        const iEnd: number = label.length;
-
-        width = 0.0;
-        while (index < iEnd && Typesetter.DELIMITERS.indexOf(label.charAt(index)) === -1) {
-            if (index > 0) {
-                width += label.kerningBefore(index);
-            }
-            width += label.advance(index);
-            ++index;
+    private static lineAnchorOffset(label: Label): number {
+        switch (label.lineAnchor) {
+            case Label.LineAnchor.Ascent:
+                return label.fontFace!.ascent;
+            case Label.LineAnchor.Center:
+                return label.fontFace!.size * 0.5 + label.fontFace!.descent;
+            case Label.LineAnchor.Descent:
+                return label.fontFace!.descent;
+            case Label.LineAnchor.Top:
+                return label.fontFace!.base;
+            case Label.LineAnchor.Bottom:
+                return label.fontFace!.base - label.fontFace!.lineHeight;
+            case Label.LineAnchor.Baseline:
+            default:
+                return 0.0;
         }
-        return index;
     }
 
+
     /**
-     * Revert advance of preceding, not depictable glyphs. Intended to be used, e.g., on line feed.
-     * The given vertical extent is increased by the label's line height and the horizontal extent is set to either
-     * 0 or the extent of the non-depictable characters.
-     * @param label - Label to compute accumulated advances for.
-     * @param index - In/out parameter: index to backtrack the advance of non-depictable glyphs.
-     * @param begin - Index to start glyph advance accumulation at.
-     * @param pen - In/out parameter: pen (typesetting position) to be used and adjusted.
-     * @param extent - In/out parameter: extent to be adjusted.
+     * Resolves a typed float array storing the advances of each of the label's characters.
```

Comment:
I am not sure if 'resolve' is the right word. Sounds like resolving a problem. Did you mean 'return' , 'compute', 'extract', ..?

Reply:
same for kernings()